src/PlanetData.js

```diff
@@ -25,49 +44,81 @@ PlanetData.prototype = {
     this.mapData.canvas.width = this.width;
     this.mapData.canvas.height = this.height;
     this.mapData.ctx = this.mapData.canvas.getContext('2d');
-    this.mapData.imageData = this.mapData.ctx.getImageData(0,0,this.mapData.canvas.width, this.mapData.canvas.height);
+    this.mapData.imageData = this.mapData.ctx.getImageData(0, 0, this.mapData.canvas.width, this.mapData.canvas.height);
     this.mapData.data = this.mapData.imageData.data;
 
-    // TODO: add hydro erosion to this
-    // TODO: add complex features, cliffs, rivers, etc.
     //create a 2d array which holds our height data for the planet
     //This is done so that we can avoid calling our noise function more than once per pixel
     var buffer = [];
-    for (var x = 0; x < this.width; x++) {
+    var rivers = [];
+    var peak = 0;
+    var low = 1;
+    var terh = 0;
+    var rv = 0;
+    for (var x = 0; x < this.mapData.canvas.width; x++) {
       buffer.push([]);
-      for (var y = 0; y < this.height; y++) {
-        buffer[x].push(noise.fbm2(x * 0.005, y * 0.005));
+      rivers.push([]);
+      for (var y = 0; y < this.mapData.canvas.height; y++) {
+        terh = this.heightmap(x * this.skew, y);
+        rv = 1 - noise.worley2(x * 25 * this.rfrequency + terh * 3, y * 25 * this.rfrequency + terh * 3);//curve the river based on height
+        rivers[x].push(rv);
+        buffer[x].push(terh - rv * 0.8 * this.erodibility * (1 - terh) * this.moisture);//decrease height around rivers
+        if (terh > peak) {
+          peak = terh;
+        } else if (terh < low) {
+          low = terh;
+        }
+        
       }
     }
-
+    //TODO: add texture to water
     //Here we will take the height data and convert it to colors
     var alpha, red, green, blue, h, c, i;
-    for (var x = 0; x < this.width - 1; x++) {
-      for (var y = 0; y < this.height - 1; y++) {
+    var waterLevel = utils.lerp(low, peak, utils.smoothstep(low, peak, this.moisture) * 0.3);
+    for (var x = 1; x < buffer.length - 1; x++) {
+      for (var y = 1; y < buffer[x].length - 1; y++) {
         h = buffer[x][y];
-        alpha = 255;
+        rv = rivers[x][y];
+        var w = 0;
+
+        //this block determines where the rivers are
+        /*todo: use an analytic derivitive to compute w rather than a logic block*/
+        if (x > 1 && y > 1) {
+          if (((rv >= rivers[x + 1][y] && rv >= rivers[x - 1][y])||(rv >= rivers[x][y + 1] && rv >= rivers[x][y - 1]))&&(rv > (0.7 - this.moisture))) {

```

Comment:
This line is quite long, I think moving it to multiple lines will make it easier to read.